### PR TITLE
Binary operations on `non-empty-string` and `string` should error

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/6624

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
The right solution here is to introduce `Type::isString()` and use that. For example your code will continue to fail with `literal-string`.

**Comment by staabm** [comment]
> The right solution here is to introduce `Type::isString()` and use that

hmm do you mean adding this method on the `Type` interface and implement it in all types?
if so, this is a BC break, isn't it?

**Comment by staabm** [comment]
I see. I got the impression because the `Type`-interface has a `@api` annotation

grafik

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -1419,6 +1424,12 @@ private function resolveType(Expr $node): Type
 
 			$resultType = TypeCombinator::union($leftNumberType, $rightNumberType);
 			if ($node instanceof Expr\AssignOp\Div || $node instanceof Expr\BinaryOp\Div) {
+				// division on strings is not allowed. its allowed on numeric strings though.
+				if ($leftType->isString()->yes() && !$leftType->isNumericString()->yes() && !$leftType->isLiteralString()->yes()
+					|| $rightType->isString()->yes() && !$rightType->isNumericString()->yes() && !$rightType->isLiteralString()->yes()) {
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1429]: 1) `||` vs. `&&` operator priority needs to be disambiguated with parentheses
2) I don't know why literal-string is mentioned in this condition

File: src/Analyser/MutatingScope.php [hunk 2]
```diff
@@ -1400,6 +1400,11 @@ private function resolveType(Expr $node): Type
 				return new ErrorType();
 			}
 
+			if ($leftType->isNonEmptyString()->yes() && !$leftType->isNumericString()->yes() ||
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:1403]: 1) `||` vs. `&&` operator priority needs to be disambiguated with parentheses
2) Why `isNonEmptyString` and not `isString`?

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (staabm) [reply src/Analyser/MutatingScope.php]: 1. fixed
2. I mixed up my mental model in "123" would be a literal-string, but it is also a numeric-string, so the additional check is not necessary. fixed a bug so we don't accept `literal-string` for binary operations

**Pull request merged**